File: 0042-trapping-rain-water/0042-trapping-rain-water.py
class Solution:
    def trap(self, height: List[int]) -> int:
        i = 0
        j = len(height) - 1
        start = 0
        end = 0
        res = 0
        while (i < j):
            start = height[i] if start < height[i] else start
            end = height[j] if end < height[j] else end
            if start <= end:
                res += (start - height[i])
                i += 1
            else:
                res += (end - height[j])
                j -= 1
        return res

File: 0042-trapping-rain-water/0042-trapping-rain-water.cpp
class Solution {
public:
    int trap(vector<int>& height) {
        int i = 0, j = height.size() - 1;
        int start = height[i], end = height[j];
        int sum = 0;
        while (i < j) {
            start = start < height[i] ? height[i] : start;
            end = end < height[j] ? height[j] : end;
            if (start < end) {
                sum += (start - height[i]);
                i++;
            } else {
                sum += (end - height[j]);
                j--;  
            }
        }
        return sum;
    }
};